packages/api/internal/dns/server.go

```diff
@@ -2,99 +2,242 @@ package dns
 
 import (
 	"context"
+	"errors"
 	"fmt"
-	"log"
 	"net"
 	"strings"
 	"sync"
+	"time"
 
+	"github.com/go-redis/cache/v8"
 	"github.com/go-redis/redis/v8"
 	resolver "github.com/miekg/dns"
+	"go.uber.org/zap"
 
 	"github.com/e2b-dev/infra/packages/shared/pkg/smap"
 )
 
 const ttl = 0
+const redisTTL = 24 * time.Hour
 
 const defaultRoutingIP = "127.0.0.1"
 
+const cachedDnsPrefix = "sandbox.dns."
+
 type DNS struct {
-	mu      sync.Mutex
-	redis   *redis.Client
-	records *smap.Map[string]
+	srv    *resolver.Server
+	logger *zap.SugaredLogger
+
+	remote *cache.Cache
+	local  *smap.Map[string]
+
+	closer struct {
+		once sync.Once
+		op   func(context.Context) error
+		err  error
+	}
 }
 
-func New(rc *redis.Client) *DNS {
-	return &DNS{
-		redis:   rc,
-		records: smap.New[string](),
+func New(ctx context.Context, rc *redis.Client, logger *zap.SugaredLogger) *DNS {
+	d := &DNS{logger: logger}
+
+	if rc != nil {
+		d.remote = cache.New(&cache.Options{Redis: rc, LocalCache: cache.NewTinyLFU(10_000, time.Hour)})
+	} else {
+		d.local = smap.New[string]()
 	}
+
+	return d
 }
 
-func (d *DNS) Add(_ context.Context, sandboxID, ip string) {
-	d.records.Insert(d.hostname(sandboxID), ip)
+func (d *DNS) Add(ctx context.Context, sandboxID, ip string) {
+	switch {
+	case d.remote != nil:
+		d.remote.Set(&cache.Item{
+			Ctx:   ctx,
+			TTL:   redisTTL,
+			Key:   sandboxID,
+			Value: ip,
+		})
+	case d.local != nil:
+		d.local.Insert(sandboxID, ip)
+	}
 }
 
-func (d *DNS) Remove(_ context.Context, sandboxID, ip string) {
-	d.records.RemoveCb(d.hostname(sandboxID), func(key string, v string, exists bool) bool {
-		return v == ip
-	})
+func (d *DNS) Remove(ctx context.Context, sandboxID, ip string) {
+	switch {
+	case d.remote != nil:
+		if err := d.remote.Delete(ctx, d.cacheKey(sandboxID)); err != nil {
+			d.logger.Debug("removing item from DNS cache", zap.Error(err), zap.String("sandbox", sandboxID))
+		}
+	case d.local != nil:
+		d.local.RemoveCb(d.cacheKey(sandboxID), func(k string, v string, ok bool) bool { return v == ip })
```

Comment:
this "check the value of the IP" before deleting things, is previous behavior. I'm not sure what it's trying to protect against, and if it's worth retaining. I think we probably should either unconditionally remove keys from the cache or add this additional check to the remote case.

Reply:
We introduce pause possibility, which can cause the same `sandboxID` is used multiple times and there were some issues with resolving the hostname. We though there's some race condition, but turned out it's nginx issue

Reply:
No need for that

Reply:
cool, it's been removed in my latest push.